bee-api/bee-rest-api/src/endpoints/permission.rs

```diff
@@ -1,30 +1,133 @@
-    allowed_ips: Box<[IpAddr]>,
+lazy_static! {
+    static ref DASHBOARD_ROUTES: RegexSet = {
+        let routes = vec![
+            "/api/v1/info",
+            "/api/v1/messages/*",
+            "/api/v1/outputs/*",
+            "/api/v1/addresses/*",
+            "/api/v1/milestones/*",
+            "/api/v1/peers*",
+        ];
+        // Panic: unwrapping is fine because all strings in `routes` can be turned into valid regular expressions.
+        RegexSet::new(routes.iter().map(|r| route_to_regex(r)).collect::<Vec<_>>()).unwrap()
+    };
+}
+
+pub(crate) fn check_permission<B: StorageBackend>(
+    args: ApiArgsFullNode<B>,
 ) -> impl Filter<Extract = (), Error = Rejection> + Clone {
-    warp::addr::remote()
-        .and_then(move |addr: Option<SocketAddr>| {
-            let route = route.to_owned();
-            let public_routes = public_routes.clone();
-            let allowed_ips = allowed_ips.clone();
+    warp::any()
+        .and(warp::filters::path::full())
+        .and(warp::header::headers_cloned())
+        .and_then(move |path: FullPath, headers| {
+            let args = args.clone();
             async move {
-                if let Some(v) = addr {
-                    if allowed_ips.contains(&v.ip()) || public_routes.contains(&route) {
+                // Check if the requested endpoint is open for public use.
+                if args.rest_api_config.public_routes().is_match(path.as_str()) {
+                    return Ok(());
+                }
+
+                // Else the requested endpoint is protected by JWT / or not even exposed.
+                let jwt = extract_jwt(&headers)?;
+                // Decode the JWT payload to find out how to validate it. The `aud` claim will indicate if it's an API
+                // JWT or a Dashboard JWT.
+                let jwt_payload = {
+                    let jwt_string = jwt.to_string();
+                    // Every JWT consists of 3 parts: 1) header, 2) payload, 3) signature.
+                    // The different parts are separated by `.`.
+                    let split = jwt_string.split('.').collect::<Vec<_>>();
+                    // If there are less or more then 3 parts the given JWT is invalid.
+                    if split.len() != 3 {
+                        return Err(reject::custom(CustomRejection::Forbidden));
+                    }
+
+                    // Get the base64 encoded payload.
+                    let encoded_payload = split[1];
+                    // Base64-decode the payload to bytes.
+                    let decoded_payload_bytes =
+                        base64::decode(encoded_payload).map_err(|_| reject::custom(CustomRejection::Forbidden))?;
+                    // Create a UTF-8 string from the decoded bytes.
+                    String::from_utf8(decoded_payload_bytes).map_err(|_| CustomRejection::Forbidden)?
+                };
+
+                if jwt_payload.contains(&format!("\"aud\":\"{}\"", API_AUDIENCE_CLAIM)) {
```

Comment:
is there any way to put the argument of `contains` into a constant to avoid computing it on every call?

Reply:
I changed it to:
```
pub const API_AUDIENCE_CLAIM: &str = "api";
pub const DASHBOARD_AUDIENCE_CLAIM: &str = "dashboard";
const API_JWT_HINT: &str = "\"aud\":\"api\"";
#[cfg(feature = "dashboard")]
const DASHBOARD_JWT_HINT: &str = "\"aud\":\"dashboard\"";
```

Reply:
Hmm not sure if this works:

```
error: expected a literal
  --> bee-api\bee-rest-api\src\endpoints\permission.rs:20:51
   |
20 | const API_JWT_HINT: &str =  concat!("\"aud\":\"", API_AUDIENCE_CLAIM, "\"");
   |                                                   ^^^^^^^^^^^^^^^^^^
   |
   = note: only literals (like `"foo"`, `42` and `3.14`) can be passed to `concat!()`
```

Reply:
Limited with the constants, so we will simply do:

```
pub const API_AUDIENCE_CLAIM: &str = "api";
pub const DASHBOARD_AUDIENCE_CLAIM: &str = "dashboard";
const API_JWT_HINT: &str = "\"aud\":\"api\"";
#[cfg(feature = "dashboard")]
const DASHBOARD_JWT_HINT: &str = "\"aud\":\"dashboard\"";
```